vllm/model_executor/parameter.py

```diff
@@ -0,0 +1,191 @@
+from typing import Callable, List, Optional, Tuple, Union
+
+import torch
+from torch.nn import Parameter
+
+from vllm.distributed import get_tensor_model_parallel_rank
+from vllm.logger import init_logger
+
+__all__ = [
+    "vLLMParameter", "PackedvLLMParameter", "PerTensorScaleParameter",
+    "ModelWeightParameter", "ChannelQuantScaleParameter",
+    "GroupQuantScaleParameter"
+]
+
+logger = init_logger(__name__)
+
+
+class vLLMParameter(Parameter):
+
+    def __new__(cls, data: torch.Tensor, **kwargs):
+
+        return super().__new__(cls, data=data, requires_grad=False)
+
+    def __init__(self,
+                 data: torch.Tensor,
+                 weight_loader: Callable,
+                 ignore_warnings: Optional[bool] = True):
+
+        self._ignore_warnings = True
+        self._weight_loader = weight_loader
+
+    @property
+    def weight_loader(self):
+        return self._weight_loader
+
+
+class _ColumnvLLMParameter(vLLMParameter):
+
+    def __init__(self, output_dim: int, **kwargs):
+        self._output_dim = output_dim
+        super().__init__(**kwargs)
+
+    @property
+    def output_dim(self):
+        return self._output_dim
+
+    def load_column_parallel_weight(self, loaded_weight: torch.Tensor,
+                                    **kwargs):
+        tp_rank = get_tensor_model_parallel_rank()
+        shard_size = self.data.shape[self.output_dim]
+        loaded_weight = loaded_weight.narrow(self.output_dim,
+                                             tp_rank * shard_size, shard_size)
+
+    def load_merged_column_weight(self, param_data: torch.Tensor,
+                                  loaded_weight: torch.Tensor,
+                                  shard_offset: int, shard_size: int,
+                                  **kwargs):
+
+        if isinstance(
+                self,
+                PackedvLLMParameter) and self.packed_dim == self.output_dim:
+            shard_size, shard_offset = self.adjust_shard_indexes_for_packing(
+                shard_offset=shard_offset, shard_size=shard_size)
+
+        tp_rank = get_tensor_model_parallel_rank()
+        param_data = param_data.narrow(self.output_dim, shard_offset,
+                                       shard_size)
+        loaded_weight = loaded_weight.narrow(self.output_dim,
+                                             tp_rank * shard_size, shard_size)
+        return param_data, loaded_weight
+
+    def load_qkv_weight(self, param_data: torch.Tensor,
+                        loaded_weight: torch.Tensor, num_heads: int,
```

Comment:
We should not have to pass `param_data` to this function since `param_data=self.data`

Reply:
yup, typo